pkg/kafka/partition/reader.go

```diff
@@ -114,7 +154,251 @@ func (p *Reader) run(ctx context.Context) error {
+	// only, which is more performant than requesting the offsets for all partitions.
+	partitionReq := kmsg.NewListOffsetsRequestTopicPartition()
+	partitionReq.Partition = p.partitionID
+	partitionReq.Timestamp = position
+
+	topicReq := kmsg.NewListOffsetsRequestTopic()
+	topicReq.Topic = p.kafkaCfg.Topic
+	topicReq.Partitions = []kmsg.ListOffsetsRequestTopicPartition{partitionReq}
+
+	req := kmsg.NewPtrListOffsetsRequest()
+	req.IsolationLevel = 0 // 0 means READ_UNCOMMITTED.
+	req.Topics = []kmsg.ListOffsetsRequestTopic{topicReq}
+
+	// Even if we share the same client, other in-flight requests are not canceled once this context is canceled
+	// (or its deadline is exceeded). We've verified it with a unit test.
+	resps := p.client.RequestSharded(ctx, req)
+
+	// Since we issued a request for only 1 partition, we expect exactly 1 response.
+	if expected := 1; len(resps) != expected {
+		return 0, fmt.Errorf("unexpected number of responses (expected: %d, got: %d)", expected, len(resps))
+	}
+
+	// Ensure no error occurred.
+	res := resps[0]
+	if res.Err != nil {
+		return 0, res.Err
+	}
+
+	// Parse the response.
+	listRes, ok := res.Resp.(*kmsg.ListOffsetsResponse)
+	if !ok {
+		return 0, errors.New("unexpected response type")
+	}
+	if expected, actual := 1, len(listRes.Topics); actual != expected {
+		return 0, fmt.Errorf("unexpected number of topics in the response (expected: %d, got: %d)", expected, actual)
+	}
+	if expected, actual := p.kafkaCfg.Topic, listRes.Topics[0].Topic; expected != actual {
+		return 0, fmt.Errorf("unexpected topic in the response (expected: %s, got: %s)", expected, actual)
+	}
+	if expected, actual := 1, len(listRes.Topics[0].Partitions); actual != expected {
+		return 0, fmt.Errorf("unexpected number of partitions in the response (expected: %d, got: %d)", expected, actual)
+	}
+	if expected, actual := p.partitionID, listRes.Topics[0].Partitions[0].Partition; actual != expected {
+		return 0, fmt.Errorf("unexpected partition in the response (expected: %d, got: %d)", expected, actual)
+	}
+	if err := kerr.ErrorForCode(listRes.Topics[0].Partitions[0].ErrorCode); err != nil {
+		return 0, err
+	}
+
+	return listRes.Topics[0].Partitions[0].Offset, nil
+}
+
+// processNextFetchesUntilTargetOrMaxLagHonored process records from Kafka until at least the maxLag is honored.
+// This function does a best-effort to get lag below targetLag, but it's not guaranteed that it will be
+// reached once this function successfully returns (only maxLag is guaranteed).
+func (p *Reader) processNextFetchesUntilTargetOrMaxLagHonored(ctx context.Context, targetLag, maxLag time.Duration, recordsChan chan<- []Record) error {
+	logger := log.With(p.logger, "target_lag", targetLag, "max_lag", maxLag)
+	level.Info(logger).Log("msg", "partition reader is starting to consume partition until target and max consumer lag is honored")
+
+	attempts := []func() (currLag time.Duration, _ error){
```

Comment:
nit: couldn't we make it like this? Or am I missing something?
```suggestion
	attempts := []func() (time.Duration, error){
```